api/client.go

```diff
@@ -0,0 +1,946 @@
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	c := &Client{c: cfg}
+	dialer := grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+		if c.isClosed() {
+			return nil, trace.ConnectionProblem(nil, "client is closed")
+		}
+		conn, err := c.c.Dialer.DialContext(ctx, "tcp", addr)
+		if err != nil {
+			return nil, trace.ConnectionProblem(err, "failed to dial")
+		}
+		return conn, nil
+	})
+
+	tlsConfig := c.c.TLS.Clone()
+	tlsConfig.NextProtos = []string{http2.NextProtoTLS}
+
+	var err error
+	if c.conn, err = grpc.Dial(teleport.APIDomain,
+		dialer,
+		grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)),
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:                c.c.KeepAlivePeriod,
+			Timeout:             c.c.KeepAlivePeriod * time.Duration(c.c.KeepAliveCount),
+			PermitWithoutStream: true,
+		}),
+	); err != nil {
+		return nil, trail.FromGRPC(err)
+	}
+
+	c.grpc = proto.NewAuthServiceClient(c.conn)
+	return c, nil
+}
+
+// Close closes the Client connection to the auth server
+func (c *Client) Close() error {
+	if !c.setClosed() {
+		return nil
+	}
+	if c.conn != nil {
+		err := c.conn.Close()
+		c.conn = nil
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (c *Client) isClosed() bool {
+	return atomic.LoadInt32(&c.closedFlag) == 1
+}
+
+// set Client closedFlag to 1 and return whether it was changed.
+func (c *Client) setClosed() bool {
+	return atomic.CompareAndSwapInt32(&c.closedFlag, 0, 1)
+}
+
+// Ping gets basic info about the auth server.
+func (c *Client) Ping(ctx context.Context) (proto.PingResponse, error) {
+	rsp, err := c.grpc.Ping(ctx, &proto.PingRequest{})
+	if err != nil {
+		return proto.PingResponse{}, trail.FromGRPC(err)
+	}
+	return *rsp, nil
+}
+
+// UpsertNode is used by SSH servers to report their presence
+// to the auth servers in form of heartbeat expiring after ttl period.
+func (c *Client) UpsertNode(s services.Server) (*services.KeepAlive, error) {
+	if s.GetNamespace() == "" {
+		return nil, trace.BadParameter("missing node namespace")
+	}
+	protoServer, ok := s.(*services.ServerV2)
+	if !ok {
+		return nil, trace.BadParameter("unsupported client")
+	}
+	keepAlive, err := c.grpc.UpsertNode(context.TODO(), protoServer)
+	if err != nil {
+		return nil, trail.FromGRPC(err)
+	}
+	return keepAlive, nil
+}
+
+// NewKeepAliver returns a new instance of keep aliver.
+// Run k.Close to release the keepAliver and its goroutines
+func (c *Client) NewKeepAliver(ctx context.Context) (services.KeepAliver, error) {
+	cancelCtx, cancel := context.WithCancel(ctx)
+	stream, err := c.grpc.SendKeepAlives(cancelCtx)
+	if err != nil {
+		cancel()
+		return nil, trail.FromGRPC(err)
+	}
+	k := &streamKeepAliver{
+		stream:      stream,
+		ctx:         cancelCtx,
+		cancel:      cancel,
+		keepAlivesC: make(chan services.KeepAlive),
+	}
+	go k.forwardKeepAlives()
+	go k.recv()
+	return k, nil
+}
+
+type streamKeepAliver struct {
```

Comment:
I'd honestly consider moving "keep aliver" and "stream watcher" (and client's respective NewXXX methods) , as well as "audit streamer" and "delegator" into separate .go files in the same package to keep things clean, they look a bit out of place here.

Reply:
do you mean individual go files, or would a `streams.go` file work well here?